drape/harfbuzz_shaping.cpp

```diff
@@ -0,0 +1,218 @@
+    }
+  }
+
+  *script = scripts[0];
+  return length;
+}
+
+// A copy of hb_icu_script_to_script to avoid direct ICU dependency.
+hb_script_t ICUScriptToHarfbuzzScript(UScriptCode script)
+{
+  if (script == USCRIPT_INVALID_CODE)
+    return HB_SCRIPT_INVALID;
+  return hb_script_from_string(uscript_getShortName (script), -1);
+}
+
+void GetSingleTextLineRuns(TextSegments & runs)
+{
+  auto const & text = runs.m_text;
+  auto const textLength = static_cast<int32_t>(text.length());
+
+  // Deliberately not checking for nullptr.
+  thread_local UBiDi * const bidi = ubidi_open();
+  UErrorCode error = U_ZERO_ERROR;
+  ::ubidi_setPara(bidi, text.data(), textLength, UBIDI_DEFAULT_LTR, nullptr, &error);
+  if (U_FAILURE(error))
+  {
+    LOG(LERROR, ("ubidi_setPara failed with code", error));
+    runs.m_segments.emplace_back(0, 0, HB_SCRIPT_UNKNOWN, HB_DIRECTION_INVALID);
+    return;
+  }
+
+  // Split the original text by logical runs, then each logical run by common script and each sequence at special
+  // characters and style boundaries. This invariant holds: bidi_run_start <= script_run_start <= breakingRunStart
+  // <= breakingRunEnd <= script_run_end <= bidi_run_end. AB: Breaking runs are dropped, they may not be needed.
+  for (int32_t bidiRunStart = 0; bidiRunStart < textLength;)
+  {
+    // Determine the longest logical run (e.g. same bidi direction) from this point.
+    int32_t bidiRunBreak = 0;
+    UBiDiLevel bidiLevel = 0;
+    ::ubidi_getLogicalRun(bidi, bidiRunStart, &bidiRunBreak, &bidiLevel);
+    int32_t const bidiRunEnd = bidiRunBreak;
+    ASSERT_LESS(bidiRunStart, bidiRunEnd, ());
+
+    for (int32_t scriptRunStart = bidiRunStart; scriptRunStart < bidiRunEnd;)
+    {
+      // Find the longest sequence of characters that have at least one common UScriptCode value.
+      UScriptCode script = USCRIPT_INVALID_CODE;
+      size_t const scriptRunEnd = ScriptInterval(runs.m_text, scriptRunStart, bidiRunEnd - scriptRunStart, &script) + scriptRunStart;
+      ASSERT_LESS(scriptRunStart, base::asserted_cast<int32_t>(scriptRunEnd), ());
+
+      // TODO(AB): May need to break on different unicode blocks, parentheses, and control chars (spaces).
+
+      // TODO(AB): Support vertical layouts if necessary.
+      runs.m_segments.emplace_back(scriptRunStart, scriptRunEnd - scriptRunStart, ICUScriptToHarfbuzzScript(script), bidiLevel & 0x01 ? HB_DIRECTION_RTL : HB_DIRECTION_LTR);
+
+      // Move to the next script sequence.
+      scriptRunStart = static_cast<int32_t>(scriptRunEnd);
+    }
+    // Move to the next direction sequence.
+    bidiRunStart = bidiRunEnd;
+  }
+}
+
+void ReorderRTL(TextSegments & runs)
+{
+  // TODO(AB): Optimize implementation to use indexes to runs instead of copying runs.
+  auto it = runs.m_segments.begin();
+  auto const end = runs.m_segments.end();
+  // TODO(AB): Line (default rendering) direction is determined by the first run. It should be defined as a parameter depending on the language.
```

Comment:
nit: Soft rule to keep 125 chars in a line for better review :) I have a visual margin in my editor.